### ISSUE TITLE
Squashing two commits should not depend on the base ref 

### ISSUE THREAD

**Issue by mnieber** [divmain/GitSavvy#619]
Squashing two consecutive commits is always technically possible (there is no chance of a merge conflict), but GitSavvy will not do it if the base ref for the rebase indicates a merge conflict.

Similarly, squashing non-consecutive commits (let's call them A and B) should not depend on the base ref in the rebase screen (it should only depend on the commits starting after A and leading up to B).

ps I love GitSavvy

**Comment by randy3k** [comment]
Currently, any commit manipulation operations [check][1] whether the current branch is rebased or not to the base ref (imagine `dev` is rebased on `master).  It would not be necessary if say, we only want to squash `B` and `C`. The merge base of squashing `B` and `C` is `B`, it should only check if `dev` is a descendent of `B`. The check code should move to [2].


```
* dev
|
* C
|
* B
|
| * master
| |
| * A
|/
|
* M
```




[1]: https://github.com/divmain/GitSavvy/blob/8cfeb94360d71e0b3f7b40ed9f130cfe10b23a90/core/interfaces/rebase.py#L465-L465
[2]: https://github.com/divmain/GitSavvy/blob/8cfeb94360d71e0b3f7b40ed9f130cfe10b23a90/core/interfaces/rebase.py#L507-L507

**Comment by randy3k** [comment]
Though I assign this to myself but I don't guarantee a prompt devliver time.

**Comment by mnieber** [comment]
I tried out rebasing with the new version today, and the problem persists. If I have the commits described below, then it should always be possible to squash E (the current HEAD) into C. The reason is that for this particular squash operation, if suffices to consider only the commits C, D and E (commits A and B are completely irrelevant). Since commit D does not touch 1.txt, the squash of the two commits touching 1.txt is guaranteed to work, but unfortunately, GitSavvy doesn't allow it.

A
B
C (changes 1.txt)
D (changes 2.txt)
E (changes 1.txt)

**Comment by randy3k** [comment]
I do expect it should work in the most recent release. (Though I am on a mobile and cannot test it now)

**Comment by mnieber** [comment]
Hi Randy, 

I tried checking my local version and the package-metadata.json in my GitSavvy.sublime-package file says version 2.14.2.

To summarize: after GitSavvy refused to squash the commits, I rebased on the command line using "git rebase --interactive <hash-of-commit-B>", I moved commit E up one position, and marked it as "squash". How is that different from what GitSavvy tries to do? Would it be possible to just let GitSavvy try the squash, and abort the rebase if it failed? (instead of trying to predict in advance whether the rebase is possible)

**Comment by randy3k** [comment]
What is the error message when it failed to squash?

**Comment by randy3k** [comment]
> Would it be possible to just let GitSavvy try the squash, and abort the rebase if it failed? (instead of trying to predict in advance whether the rebase is possible)

This is the current behavior.

**Comment by mnieber** [comment]
>>What is the error message when it failed to squash?
I will let you know the next time I see it :-).

**Comment by stoivo** [comment]
I personally love the way git rebase -i work so I wonder if we should add this custom command to the default installation. 
```json
    {
        "caption": "git: rebase interactive (rebase -i)",
        "command": "gs_custom",
        "args": {
            "args": ["rebase", "-i", "{PROMPT_ARG}"],
            "prompt_msg": "Rebase from",
            "start_msg": "Starting rebase interactive...",
            "complete_msg": "Rebase interactive started",
            "custom_environ": {
                "EDITOR": "/usr/local/bin/subl -nw"
            },
        }
    },
```
I am only using this one. 
When using it you will need to return to the command line if it fails. You will need to run 'git rebase --continue'

demo:

![ezgif-3-21530a8397](https://user-images.githubusercontent.com/3492040/27001544-c67e54fe-4dcc-11e7-8519-328795158147.gif)

**Comment by randy3k** [comment]
@mnieber 

I just test on my computer locally. It works as expected. 

Try checking out the [subfeature](https://github.com/randy3k/RebaseTest/tree/subfeature) branch of https://github.com/randy3k/RebaseTest
and rebase G2 to G.

**Comment by randy3k** [comment]
@stoivo, unfortunately, `rebase -i` doesn't preserve merge.  Though there is an option `-p` to preserve merge in `rebase` command, it is not recommended to be used with the `-i` flag.

> This uses the --interactive machinery internally, but combining it with the --interactive option explicitly is generally not a good idea unless you know what you are doing (see BUGS below).

source: https://git-scm.com/docs/git-rebase

**Comment by mnieber** [comment]
@randy3k I just got the error again when trying to squash the last two commits in my  branch (which in theory should always work), it says: "Unable to manipulate commits while repo is in unclean state".

I don't think my repo is in an unclean state, since I only have a few untracked files. The output of `git status` is:

On branch master
Your branch is ahead of 'origin/master' by 8 commits.
  (use "git push" to publish your local commits)
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	django/.cache/
	django/db.sqlite3

nothing added to commit but untracked files present (use "git add" to track)

**Comment by randy3k** [comment]
Having untracked files is consider being unclean in the moment.
#644 is an attempt to fix it.

**Comment by mnieber** [comment]
@randy3k 

The problem also occurs when there are no untracked files.

I still don't understand why GitSavvy doesn't follow the workflow that I do manually, which is:
- start an interactive rebase
- move the commit to it's new location, and mark it as "fixup"
- exit the interactive rebase
- if git tells me that the rebase fails, execute `git rebase --abort`

If this workflow would be executed automatically when the "[Q] squash commit with ..." command is run, then I would not have to do this operation manually 50% of the time. I don't want to sound like I am complaining, I love GitSavvy, and I would love to use it also for squashing commits (I'm talking about the cases where squashing the commit manually produces no conflicts, which suggests that an automatic procedure should also have no problems)

**Comment by randy3k** [comment]
That is unexpected. Do you have any more details about the git status. Btw, did you also try `git: fix up from stage`

**Comment by randy3k** [comment]
It would be great if you can open a new issue and give more details about your issue as this thread is considered fixed.

**Comment by mnieber** [comment]
The status is `nothing to commit, working directory clean`. 

I just made an interesting observation. First, I found that I could not drop the last commit in my branch (using `[p] drop commit`). Then, I changed the rebase ref using `[f] define base ref for dashboard` to be the predecessor of the last commit. Now, I could drop the last commit. So it does seem that an operation which should always succeed may fail in practice depending on the rebase base ref.
